application/common/dos.py

```diff
@@ -209,22 +137,23 @@ def get_all_valid_dos_postcodes() -> Set[str]:
     Returns: A set of normalised postcodes as strings"""
     logger.info("Collecting all valid postcodes from DoS DB")
     sql_command = "SELECT postcode FROM locations"
-    c = query_dos_db(sql_command)
-    postcodes = set(row["postcode"].replace(" ", "").upper() for row in c.fetchall())
+    with connect_to_dos_db_replica() as connection:
+        cursor = query_dos_db(connection=connection, query=sql_command)
+        postcodes = set(row["postcode"].replace(" ", "").upper() for row in cursor.fetchall())
+        cursor.close()
     logger.info(f"Found {len(postcodes)} unique postcodes from DoS DB.")
     return postcodes
 
 
 def get_valid_dos_postcode(postcode: str) -> Union[str, None]:
     """Finds the valid DoS formatted version of the given postcode. Or None if not a valid DoS postcode"""
     dos_locations = [loc for loc in get_dos_locations(postcode) if loc.is_valid()]
-    if len(dos_locations) == 0:
+    if not dos_locations:
         return None
     return dos_locations[0].postcode
 
 
 def get_services_from_db(typeids: Iterable) -> List[DoSService]:
```

Comment:
This function needs reworking to only use one sql query. The way it uses 'IN' statements instead of 'JOIN's for large datasets can be quite taxing on the database.

If it's too big of an issue to handle as part of this ticket we can do it as a maintance task with high priority

Reply:
A ticket has been added for this (DI-554)